### PR TITLE
Enable BayesOpt

### PR DESCRIPTION
**Pull request by wiseodd**
Changes:

* Remove `detach()` in GLM predictive so that one can backprop from `f(x)` to `x`, in particular for optimizing acquisition functions.
* Add `functional_covariance` method for GLM predictive that returns the `(nk,)` mean and the `(nk, nk)` covariance over `n` input points and `k` outputs.
* Add example implementation of a linearized-Laplace-based BoTorch model for BayesOpt.

### CODE REVIEW

**Comment by AlexImmer** [comment]
Looks good to me. A few comments/questions:
- Could you add a small test case where the predictives are currently tested but with the setting `joint=True` asserting 1) equal marginal predictive and 2) right shape?
- To differentiate the predictive uncertainty as well, which depends on Jacobians, the Jacobians would have to be differentiable. Don't you need that? This would require `create_graph=True` apart from `retain_graph=True`.
- How well does the Laplace model work vs. the GP on these toy problems?

File: laplace/baselaplace.py [hunk 1]
```diff
@@ -628,10 +636,16 @@ def predictive_samples(self, x, pred_type='glm', n_samples=100,
...
 
     @torch.enable_grad()
-    def _glm_predictive_distribution(self, X):
+    def _glm_predictive_distribution(self, X, joint=False):
         Js, f_mu = self.backend.jacobians(X)
-        f_var = self.functional_variance(Js)
-        return f_mu.detach(), f_var.detach()
+
+        if joint:
+            f_mu = f_mu.flatten()  # (batch*out)
```
Comment (AlexImmer) [inline laplace/baselaplace.py:643]: Could this be risky when we have multi-output regression or do we not support this anyway?

File: laplace/curvature/backpack.py [hunk 2]
```diff
@@ -41,7 +41,7 @@ def jacobians(self, x):
                 if model.output_size > 1:
                     out[:, i].sum().backward()
                 else:
-                    out.sum().backward()
+                    out.sum().backward(retain_graph=True)
```
Comment (AlexImmer) [inline laplace/curvature/backpack.py:44]: Here potentially `create_graph=True` if pred uct. needs to be differentiable. We should, at least in the future, consider this to be a backend option to turn on and off since it costs more memory to have these set to true.

**Review by AlexImmer — APPROVED** [review]

File: examples/bayesopt_example.py [hunk 3]
```diff
@@ -0,0 +1,313 @@
+import warnings
+warnings.filterwarnings('ignore')
+
+import numpy as np
+import functorch as ft
+import torch
+from torch import nn, optim
+from torch import distributions as dists
+from torch.nn import functional as F
+from gpytorch import distributions as gdists
```
Comment (runame) [inline examples/bayesopt_example.py:10]: Since there are multiple packages used here which we don't require for the laplace package itself, we might want to add an additional `requirements.txt` file or similar for the example scripts.

File: examples/regression_example.py [hunk 4]
```diff
@@ -47,9 +47,22 @@ def get_model():
...
+# 1.) Marginal predictive distribution N(f_map(x_i), var(x_i))
+# The mean is (m,k), the var is (m,k,k)
+f_mu, f_var = la(X_test)  
+
+# 2.) Joint pred. dist. N((f_map(x_1),...,f_map(x_m)), Cov(f(x_1),...,f(x_m)))
+# The mean is (m*k,) where k is the output dim. The cov is (m*k,m*k)
+f_mu_joint, f_cov = la(X_test, joint=True)  
+
+# Both should be true
+print(torch.allclose(f_mu.flatten(), f_mu_joint))
```
Comment (runame) [inline examples/regression_example.py:61]: I think using `assert` here is nicer.

File: setup.cfg [hunk 5]
```diff
@@ -38,7 +38,7 @@ install_requires =
     backpack-for-pytorch
     asdfghjkl
 # Require a specific Python version, e.g. Python 2.7 or >= 3.4
-python_requires = >=3.8
+python_requires = >=3.7
```
Comment (runame) [inline setup.cfg:41]: Any particular reason for the downgrade?

File: laplace/curvature/asdl.py [hunk 6]
```diff
@@ -63,7 +63,7 @@ def gradients(self, x, y):
         Gs : torch.Tensor
             gradients `(batch, parameters)`
         """
-        f = batch_gradient(self.model, self.lossfunc, x, y).detach()
+        f = batch_gradient(self.model, self.lossfunc, x, y)
```
Comment (runame) [inline laplace/curvature/asdl.py:66]: Does the `detach()` also have to be removed from for the Jacobians [here](https://github.com/AlexImmer/Laplace/blob/main/laplace/curvature/asdl.py#L42)?

File: laplace/curvature/backpack.py [hunk 7]
```diff
@@ -41,7 +41,7 @@ def jacobians(self, x):
...
                 to_cat = []
                 for param in model.parameters():
                     to_cat.append(param.grad_batch.detach().reshape(x.shape[0], -1))
```
Comment (runame) [inline laplace/curvature/backpack.py:47]: This `detach()` is fine?

**Review by runame — APPROVED** [review]
LGTM!

Are there cases where you might want to differentiate through NN samples, i.e. [here](https://github.com/AlexImmer/Laplace/blob/main/laplace/baselaplace.py#L640)? Also, what about `LowRankLaplace` and `LLLaplace`, see [here](https://github.com/AlexImmer/Laplace/blob/main/laplace/baselaplace.py#L923) and [here](https://github.com/AlexImmer/Laplace/blob/main/laplace/lllaplace.py#L114)?

File: laplace/curvature/backpack.py (hunk 2 above)
Comment (wiseodd) [reply laplace/curvature/backpack.py]: Fixed [bc50025](https://github.com/AlexImmer/Laplace/pull/120/commits/bc50025d079abae01a65ed1dce8c3e3fe0975e6b)

File: setup.cfg (hunk 5 above)
Comment (wiseodd) [reply setup.cfg:41]: Many people I talked to still have TF v1.x as a dependency for various reasons. The last TF v1.x doesn't support Py > 3.7.

File: laplace/curvature/backpack.py (hunk 7 above)
Comment (wiseodd) [reply laplace/curvature/backpack.py]: Fixed in [bc50025](https://github.com/AlexImmer/Laplace/pull/120/commits/bc50025d079abae01a65ed1dce8c3e3fe0975e6b)

File: laplace/curvature/asdl.py (hunk 6 above)
Comment (wiseodd) [reply laplace/curvature/asdl.py]: Fixed in [6e50ab4](https://github.com/AlexImmer/Laplace/pull/120/commits/6e50ab4e774b2834b11aa5e984bdd1c069255b49)

**Comment by wiseodd** [comment]
@AlexImmer 

> Could you add a small test case where the predictives are currently tested but with the setting `joint=True` asserting 1) equal marginal predictive and 2) right shape?
> * To differentiate the predictive uncertainty as well, which depends on Jacobians, the Jacobians would have to be differentiable. Don't you need that? This would require `create_graph=True` apart from `retain_graph=True`.

Done in [bc50025](https://github.com/AlexImmer/Laplace/pull/120/commits/bc50025d079abae01a65ed1dce8c3e3fe0975e6b).

> * How well does the Laplace model work vs. the GP on these toy problems?

Seems to work reasonably:

```
agustinuskristiadi@akmbp ~/Projects/Laplace                                                               [13:37:28]
(base) > $ python examples/bayesopt_example.py --exp_len 50 --test_func branin --acqf qEI                [±detach ●]

Test Function: branin
-------------------------------------

[BNN-LA, MSE = 85.071; Best f(x) = 0.846, curr f(x) = 3.055]: 100%|██████████████████| 50/50 [01:58<00:00,  2.37s/it]
[GP, MSE = 2081.004; Best f(x) = 0.510, curr f(x) = 12.377]: 100%|███████████████████| 50/50 [00:13<00:00,  3.83it/s]
                                                                                                                     
agustinuskristiadi@akmbp ~/Projects/Laplace                                                               [13:39:57]
(base) > $ python examples/bayesopt_example.py --exp_len 50 --test_func branin --acqf EI                 [±detach ●]

Test Function: branin
-------------------------------------

[BNN-LA, MSE = 183.844; Best f(x) = 0.939, curr f(x) = 125.097]: 100%|███████████████| 50/50 [00:59<00:00,  1.19s/it]
[GP, MSE = 4422.465; Best f(x) = 1.199, curr f(x) = 2.495]: 100%|████████████████████| 50/50 [00:06<00:00,  8.30it/s]
```

@runame 

> Are there cases where you might want to differentiate through NN samples? 

There is, e.g. in Thompson sampling like https://arxiv.org/abs/1706.01825. I added the option.

> Also, what about LowRankLaplace

There's already `functional_covariance` function for `LowRankLaplace` here https://github.com/AlexImmer/Laplace/blob/6e50ab4e774b2834b11aa5e984bdd1c069255b49/laplace/baselaplace.py#L1016. Do I need to do something else?

> LLLaplace?

[f0981a7](https://github.com/AlexImmer/Laplace/pull/120/commits/f0981a733b52adb9f0a52e322102c935a6b0765a)

**Comment by wiseodd** [comment]
@AlexImmer, @runame: For making the last-layer Jacobians optionally backpropable, I need to do the following. I'm not sure how to make `forward_with_features` communicate with the hook, though. 

```python
def forward_with_features(self, x: torch.Tensor, enable_backprop=True)
    """
    Add an option `enable_backprop`
    """
    pass

def _get_hook(self, name: str) -> Callable:
    def hook(_, input, __):
        # Depending on the above `enable_backprop`, we detach
        self._features[name] = input[0].detach()
    return hook
```

Should I `enable_backprop` a class-level attribute? (In this case I also need to do the same with `BackPackInterface`.)

**Comment by runame** [comment]
> > Also, what about LowRankLaplace
> 
> There's already functional_covariance function for LowRankLaplace here
> 
> [Laplace/laplace/baselaplace.py](https://github.com/AlexImmer/Laplace/blob/6e50ab4e774b2834b11aa5e984bdd1c069255b49/laplace/baselaplace.py#L1016)
> 
> Line 1016 in [6e50ab4](https://github.com/AlexImmer/Laplace/commit/6e50ab4e774b2834b11aa5e984bdd1c069255b49)
>  def functional_covariance(self, Jacs): 
> . Do I need to do something else?

I was asking because of [this line](https://github.com/AlexImmer/Laplace/blob/main/laplace/baselaplace.py#L923): you detached `self.mean` for all `ParametricLaplace` subclasses, but not for LowRankLaplace.

**Comment by runame** [comment]
> For making the last-layer Jacobians optionally backpropable, I need to do the following. (...)

Yeah I think a class-level attribute is probably the best solution. Also, maybe it makes sense to consistently name the argument which decides wether to detach or not; in some places it's called `detach` and here `enable_backprop`.

File: laplace/lllaplace.py [hunk 8]
```diff
@@ -113,16 +113,23 @@ def fit(self, train_loader, override=True):
         super().fit(train_loader, override=override)
         self.mean = parameters_to_vector(self.model.last_layer.parameters()).detach()
```
Comment (runame) [inline laplace/lllaplace.py:114]: @wiseodd Does the `detach()` have to stay here? You removed it for `ParametricLaplace` subclasses (same question as for `LowRankLaplace`).

File: examples/bayesopt_example.py (hunk 3 above)
Comment (wiseodd) [reply examples/bayesopt_example.py:10]: [7726b3d](https://github.com/AlexImmer/Laplace/pull/120/commits/7726b3d9ee431e2a6c258ab27ddecd3b60d17b56)

File: examples/regression_example.py (hunk 4 above)
Comment (wiseodd) [reply examples/regression_example.py]: [7726b3d](https://github.com/AlexImmer/Laplace/pull/120/commits/7726b3d9ee431e2a6c258ab27ddecd3b60d17b56)

**Comment by wiseodd** [comment]
@AlexImmer, @runame how do you actually enable backprop in ASDL's jacobian [here](https://github.com/AlexImmer/Laplace/blob/ada1c6f9a4aa879939de52c8fa454ff27cbfe5a4/laplace/curvature/asdl.py#L21)? (I'm unfamiliar with its API.)

File: laplace/lllaplace.py (hunk 8 above)
Comment (wiseodd) [reply laplace/lllaplace.py]: > Yeah I think a class-level attribute is probably the best solution. Also, maybe it makes sense to consistently name the argument which decides wether to detach or not; in some places it's called detach and here enable_backprop.

See [7726b3d](https://github.com/AlexImmer/Laplace/pull/120/commits/7726b3d9ee431e2a6c258ab27ddecd3b60d17b56)

**Comment by wiseodd** [comment]
> Yeah I think a class-level attribute is probably the best solution. Also, maybe it makes sense to consistently name the argument which decides wether to detach or not; in some places it's called detach and here enable_backprop.

See [7726b3d](https://github.com/AlexImmer/Laplace/pull/120/commits/7726b3d9ee431e2a6c258ab27ddecd3b60d17b56)

File: laplace/lllaplace.py (hunk 8 above)
Comment (wiseodd) [reply laplace/lllaplace.py]: I'll update it once I know how to enable backdrop in ASDL since `LowRankLaplace` only supports ASDL.

**Comment by wiseodd** [comment]
@AlexImmer @runame: There's no more action for me in this PR, so you can review it again.
 
As for enabling all these stuffs in the ASDL backend, how about we invite Michael Aerni to contribute? If he has the code already, I can also do it---in this case, he simply needs to push his changes to a new branch. In any case, let's do this in a separate PR.

File: laplace/baselaplace.py [hunk 9]
```diff
@@ -522,7 +538,7 @@ def __call__(self, x, pred_type='glm', link_approx='probit', n_samples=100,
             Only works for `pred_type='glm'` and `link_approx='mc'`.
 
         generator : torch.Generator, optional
-            random number generator to control the samples (if sampling used)
+            random number generator to control the samples (if sampling used)=
```
Comment (runame) [inline laplace/baselaplace.py:541]: Typo?

File: laplace/lllaplace.py [hunk 10]
```diff
@@ -175,11 +192,11 @@ class KronLLLaplace(LLLaplace, KronLaplace):
     _key = ('last_layer', 'kron')
 
     def __init__(self, model, likelihood, sigma_noise=1., prior_precision=1.,
-                 prior_mean=0., temperature=1., backend=None, last_layer_name=None,
+                 prior_mean=0., temperature=1., enable_backprop=True, backend=None, last_layer_name=None,
```
Comment (runame) [inline laplace/lllaplace.py:195]: Why does `enable_backprop` defaults to `True` here?

**Review by runame — APPROVED** [review]
Let's just merge this for now.
